### PR TITLE
Merkle-Radix SQL operations for postgres DB's

### PR DESCRIPTION
**Pull request by peterschwarz**

### CODE REVIEW

File: libtransact/src/state/merkle/sql/backend/postgres.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    use crate::state::merkle::sql::migration;
+
+    lazy_static! {
+        static ref SERIAL_TEST_LOCK: Arc<Mutex<()>> = Arc::new(Mutex::new(()));
+    }
+
+    /// Execute a test against a postgres database.
+    ///
+    /// This function will run the migrations agains the database provided by the environment
+    /// variable STATE_MERKLE_SQL_POSTGRES_TEST_URL.  AFter the test completes, it will revert the
```
Comment (agunde406) [inline libtransact/src/state/merkle/sql/backend/postgres.rs:42]: AFter -> After

**Review by agunde406 — CHANGES_REQUESTED** [review]

File: libtransact/src/state/merkle/sql/operations/get_leaves.rs [hunk 2]
```diff
@@ -193,16 +193,102 @@ mod test {
...
     }
 
+    /// This tests that a leaf changed across several state root hashes is returned when requested
+    /// at the given state root hash. It verifies that:
+    /// 1. No leaves are returned for the initial state root hash
+    /// 2. The first leaf change is returned for first state root.
+    /// 3. The second leaf change is returned for the second state root.
+    #[cfg(feature = "state-merkle-sql-postgres-tests")]
+    #[test]
+    fn postgres_get_leaves_at_state_root() -> Result<(), Box<dyn std::error::Error>> {
```
[the remark this answers is not in the document]
Comment (peterschwarz) [reply libtransact/src/state/merkle/sql/operations/get_leaves.rs:209]: Fixed

File: libtransact/src/state/merkle/sql/operations/get_leaves.rs [hunk 3]
```diff
@@ -258,16 +344,85 @@ mod test {
...
     }
 
+    /// This tests that a leaf inserted then deleted across several state root hashes is returned
+    /// only on the state root hash where it exists.  It verifies that:
+    /// 1. No leaves are returned for the initial state root hash
+    /// 2. The leaf is returned for first state root.
+    /// 3. The leaf is no longer returned for the second state root.
+    #[cfg(feature = "state-merkle-sql-postgres-tests")]
+    #[test]
+    fn postgres_get_leaves_with_deletion() -> Result<(), Box<dyn std::error::Error>> {
```
[the remark this answers is not in the document]
Comment (peterschwarz) [reply libtransact/src/state/merkle/sql/operations/get_leaves.rs:357]: Fixed

File: libtransact/src/state/merkle/sql/backend/postgres.rs (hunk 1 above)
Comment (peterschwarz) [reply libtransact/src/state/merkle/sql/backend/postgres.rs]: Fixed

File: libtransact/src/state/merkle/sql/migration/postgres/mod.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -----------------------------------------------------------------------------
+ */
+
+//! Defines methods and utilities to interact with biome tables in a SQLite database.
```
Comment (agunde406) [inline libtransact/src/state/merkle/sql/migration/postgres/mod.rs:18]: Copypasta: this file mentions biome

File: libtransact/src/state/merkle/sql/operations/update_index.rs [hunk 5]
```diff
@@ -379,4 +476,269 @@ mod tests {
 
         Ok(())
     }
+
+    /// This test indexes a leaf and its tree nodes into the index as in an initial entry into the
+    /// merkle tree.  That is, there was no prior entry at that address. It
```
Comment (agunde406) [inline libtransact/src/state/merkle/sql/operations/update_index.rs:481]: Remove "It" at the end of this line

Comment (agunde406) [reply libtransact/src/state/merkle/sql/operations/update_index.rs:481]: Or add "verifies that"

**Review by agunde406 — CHANGES_REQUESTED** [review]

**Review by agunde406 — APPROVED** [review]

File: libtransact/src/state/merkle/sql/operations/update_index.rs [hunk 6]
```diff
@@ -148,10 +148,10 @@ impl<'a> MerkleRadixUpdateIndexOperation for MerkleRadixOperations<'a, SqliteCon
 
 #[cfg(feature = "sqlite")]
```
Comment (shannynalayna) [inline libtransact/src/state/merkle/sql/operations/update_index.rs]: Should this feature guard be removed in this commit?

Comment (shannynalayna) [reply libtransact/src/state/merkle/sql/operations/update_index.rs]: Looks like it's currently removed here: https://github.com/hyperledger/transact/pull/166/commits/17769ad91673c301afe1beeb72426f675dd62ee6

Comment (peterschwarz) [reply libtransact/src/state/merkle/sql/operations/update_index.rs]: Moved is to the appropriate commit

File: libtransact/src/state/merkle/sql/migration/postgres/mod.rs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+embed_migrations!("./src/state/merkle/sql/migration/postgres/migrations");
+
+use crate::error::InternalError;
+
+/// Run database migrations to create tables defined for the SqlMerkleState.
+///
+/// # Arguments
+///
+/// * `conn` - Connection to SQLite database
```
[the remark this answers is not in the document]
Comment (peterschwarz) [reply libtransact/src/state/merkle/sql/migration/postgres/mod.rs]: Fixed

**Review by agunde406 — APPROVED** [review]

**Review by shannynalayna — APPROVED** [review]

**Pull request merged**